### PR TITLE
[REG-1579] add initial Automated Testing package docs

### PR DESCRIPTION
**Pull request by analogrelay**
This is an initial draft of the "one-pager" Getting Started doc for the automated testing package (I think "Monkey Testing" is a little too obscure a name for this, so I was focusing on "Automated Testing" as an alternate).

TODO:

* [x] Merge a PR to RegressionGames.Unity.Testing with a few small UI improvements I made while walking through this process ([REG-1587](https://linear.app/regression-games/issue/REG-1587/[monkey-testing]-general-tidy-up-and-little-things)) - This doesn't _have_ to be done before we merge this, but the docs will refer to things I've already fixed in my branch :). https://github.com/Regression-Games/RegressionGames.Unity.Testing/pull/6
* [x] Add screenshots
* [x] Add GIF recording of Monkey bot
* [ ] Release a `v0.0.1` tag of the `RegressionGames.Unity.Testing` package on public GitHub. - This also doesn't _have_ to be done before merge but the docs will refer to it as if it exists.

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: docs/automated-testing/getting-started.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+## Installing the package
+
+Our package is available at [https://github.com/Regression-Games/RegressionGames.Unity.Testing](https://github.com/Regression-Games/RegressionGames.Unity.Testing).
+Import it into your game by clicking **Window > Package Manager** in the Unity menu,
+and then clicking the **+** button in the top left corner of the Package Manager window.
+Select **Add package from git URL** and paste in the following URL:
+
+```
+https://github.com/Regression-Games/RegressionGames.Unity.Testing.git?path=src/gg.regression.unity.testing#v0.0.1
```
Comment (analogrelay) [inline docs/automated-testing/getting-started.md:33]: For testing the doc, I currently recommend using this Git URL instead:

```
https://github.com/Regression-Games/RegressionGames.Unity.Testing.git?path=src/gg.regression.unity.testing#analogrelay/reg-1587
```

When https://github.com/Regression-Games/RegressionGames.Unity.Testing/pull/6 lands, you could also use:

```
https://github.com/Regression-Games/RegressionGames.Unity.Testing.git?path=src/gg.regression.unity.testing
```

To get the latest `main` version

File: docs/automated-testing/getting-started.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+## Adding Entity Discoverers to your scene
+
+In order to automate entities in your game, the Automation Controller needs to be able to find them.
+The Automation Controller uses Entity Discoverers to find entities in the scene that can be automated.
+The package provides a few built-in Entity Discoverers, but you can also create your own.
+To add one of the built-in discoverers, open the **GameObject > Regression Games > Discovery** menu and select the discoverer you want to add:
+
+![The "UI Element Discoverer" option in the Regression Games submenu of the GameObject menu](img/getting-started/add-uielement-discoverer.png)
+
+As long as you have an Automation Controller in your scene, the Entity Discoverer will be **automatically** be added as a child of the Automation Controller, no matter what object is selected in the Hierarchy window.
```
Comment (abeizer) [inline docs/automated-testing/getting-started.md:66]:
```suggestion
As long as you have an Automation Controller in your scene, the Entity Discoverer will **automatically** be added as a child of the Automation Controller, no matter what object is selected in the Hierarchy window.
```

Comment (vontell) [inline docs/automated-testing/getting-started.md:66]:
```suggestion
As long as you have an Automation Controller in your scene, the Entity Discoverer will **automatically** be added as a child of the Automation Controller, no matter what object is selected in the Hierarchy window.
```

**Review by abeizer — CHANGES_REQUESTED** [review]

**Review by vontell — CHANGES_REQUESTED** [review]
Just did an initial read through, I need to actually go and complete the tutorial now. Will provide another update once I do that!

File: docs/automated-testing/getting-started.md (hunk 2 above)
Comment (abeizer) [inline docs/automated-testing/getting-started.md:66]: I submitted an additional comment in my review but it didn't make it into the review?? Reposting...

I can't seem to add an Entity Discovery even after adding my Automation Controller. I get an error message about there being no Automation Controller in the scene.

https://www.loom.com/share/eea0cf42fb4a4da9bf3769b5ed7052ba?sid=42885d93-71c7-4527-9736-13a595c6170c

Comment (analogrelay) [reply docs/automated-testing/getting-started.md:66]: Looks like the scripts aren't properly loaded (you can see the missing script warnings in the editor). That would cause us to fail to recognize the Automation Controller. Now, the bigger question is why those scripts weren't loaded 🤔

Comment (analogrelay) [reply docs/automated-testing/getting-started.md:66]: If you have a chance, could you check the Console for any compilation errors? I notice you're doing this in RGBossRoom, which already has the existing RG SDK installed. It's possible there's a conflict between them. I haven't tested the two together yet (I intended to do so, but since we're focusing on new customers it wasn't the top priority).

Comment (abeizer) [reply docs/automated-testing/getting-started.md:66]: I'm not seeing any console errors. I tried removing the package, adding it back to the project, and adding a new Automation Controller and the console gives me nada. It looks like the prefab is missing scripts when the package is added.

**Review by vontell — CHANGES_REQUESTED** [review]
Nice!! Was so easy to get this up and running. Like how quick it is to get started.

Mostly comments about clarifying certain points, and just a few small issues I ran into. Also, I notice that the logo is still not working on the overlay?

File: docs/automated-testing/getting-started.md (hunk 1 above)
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:33]: Ohhh, I keep forgetting I have this in my global git config 🤣 

```
[url "git@github.com:"]
	insteadOf = https://github.com
```

(It forces any `https://github.com` to become `git@github.com:`, so HTTPS URLs automatically use SSH instead)

Good call, I'll post an updated URL soon.

The doc is still fine though, since when the package goes live, the HTTPS URL will work because the repo will be public.

Comment (analogrelay) [reply docs/automated-testing/getting-started.md:33]: After resolving some issues @abeizer found, please use _this URL_ instead:

```
https://github.com/Regression-Games/RegressionGames.Unity.Testing.git?path=src/gg.regression.unity.testing#ashley/reg-1596
```

This is necessary until https://github.com/Regression-Games/RegressionGames.Unity.Testing/pull/8 lands

**Comment by analogrelay** [comment]
> Also, I notice that the logo is still not working on the overlay?

Fixed if you use the `ashley/reg-1596` branch (see https://github.com/Regression-Games/RegressionGames.Unity.Testing/pull/8)

File: docs/automated-testing/getting-started.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+## Adding the Automation Controller and components to your scene
+
+Once the package is installed, you need to add an Automation Controller to the scene you wish to automate.
+The Automation Controller is the hub for all automation in a scene.
+Bots are spawned as children of the Automation Controller,
+and Entity Discoverers (see below) should be added as children of the Automation Controller.
+The Controller also provides APIs for spawning bots and managing recordings (via a separate, Automation Recorder component attached to the same object).
+
+To add the Automation Controller to your scene, click **GameObject > Regression Games > Automation Controller** in the Unity menu.
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:49]: Ah, I meant the main menu bar, but right clicking also provides the same menu. I'll clarify!

File: docs/automated-testing/getting-started.md (hunk 1 above)
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:33]: > I have not http config for git, and even for public repos, it will fail.

We can provide the SSH URL too, but [currently we only provide the HTTPS link in our docs for the existing SDK](https://docs.regression.gg/quickstart#step-2-import-the-regression-games-unity-bots-sdk)

File: docs/automated-testing/getting-started.md (hunk 2 above)
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:66]: This should be fixed by https://github.com/Regression-Games/RegressionGames.Unity.Testing/pull/8 . Feel free to wait for that or use the branch directly to unblock your testing!

File: docs/automated-testing/getting-started.md (hunk 1 above)
Comment (vontell) [reply docs/automated-testing/getting-started.md:33]: Oh my apologies! I confused myself and it was actually the reverse I experienced, duh... we need to use the https one and not the ssh one, THAT was the issue I originally had. So yes we can just keep the HTTPS one

File: docs/automated-testing/getting-started.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
+---
+sidebar_label: Getting Started
+---
+
+# Getting Started with Automated Testing
+
+This quick walk-through will show you how to integrate the Regression Games Automated Testing package into your game, 
+and how to use the built-in "Monkey Bot",
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/automated-testing/getting-started.md:8]: I'm not sold on calling this a walkthrough or limiting it to the scope of the Monkey Testing bot. This page isn't in the tutorials section of the docs site, so I'd expect it to encompass every available feature of the Automated Testing package - for example, the manual recording option (which I don't see documented yet)

**Review by abeizer — CHANGES_REQUESTED** [review]
It works! Things I really like about this:
* Super easy to understand and get the bot running
* Grouping concepts/functionality within + under the Automation Controller. If I know where my controller is, I know how to access important settings and data like discoverers and recordings.
* Only capturing state + screenshot when it's relevant (not blowing up the .zip size)

One thing I don't see mentioned here is the "Destroy" option after you stop a bot -> It looks like this is supposed to be an easy way to delete recordings? It doesn't seem to be deleting my .zip files though (bug?) :-( I think its usefulness is also limited right now because if I stop my game and restart it, I don't see any previous bots in the overlay. This could be nice for a future task?

I have a mix of blocking and non-blocking comments.

File: docs/automated-testing/getting-started.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/automated-testing/getting-started.md:8]: Eh. Maybe that's fine for now until there are more bots/features to document

File: docs/automated-testing/getting-started.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
+---
+sidebar_label: Getting Started
+---
+
+# Getting Started with Automated Testing
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:5]: 👍🏻

File: docs/automated-testing/getting-started.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:8]: > Our package's Monkey Bot will mimic real user behavior and record the state of your game as it performs its testing.

I like the rewrite but might adjust this sentence a bit. I don't want to mislead users into thinking the bot will behave like a user. It certainly doesn't, it's quite random and I think we should be clear about that for now. As we integrate LLMs, it can get smarter and we can update the docs.

[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:8]: My suggestion:

> This quick walkthrough will teach you how integrate the Regression Games Automated Testing package into your game and leverage a built-in bot to perform Monkey Testing. Monkey Testing is a strategy where a user randomly interacts with UI elements to uncover potential edge-cases. The built-in "Monkey Bot" will allow you to do some basic UI hardening to ensure your game doesn't crash or behave unexpectedly when a user clicks on random UI elements.

File: docs/automated-testing/getting-started.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+The Controller also provides APIs for spawning bots and managing recordings (via a separate, Automation Recorder component attached to the same object).
+
+To add the Automation Controller to your scene, click **GameObject > Regression Games > Automation Controller** in the Unity menu.
+
+![The "Automation Controller" option in the Regression Games submenu of the GameObject menu](img/getting-started/add-automation-controller-menu.png)
+
+**By default**, the Automation Controller and all it's children will mark themselves as [DontDestroyOnLoad](https://docs.unity3d.com/ScriptReference/Object.DontDestroyOnLoad.html). This ensures that bots continue to run when you load a new scene in your game. You can disable this behavior by unchecking the "Dont Destroy On Load" checkbox on the Automation Controller component.
+
+![The Automation Controller component with the "Dont Destroy On Load" checkbox checked](img/getting-started/ac-dont-destroy-on-load.png)
+
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:56]: Not easily, no. I looked for exactly that 😞 . You can create a custom editor, which we do for Automation Recorder, but then you have to put an editor in for every property yourself.

[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:56]: This only shows in the alt-text for the image though, which softens the blow of the bad punctuation :)

[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:56]: I might just do that though...

File: docs/automated-testing/getting-started.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+## Adding the Overlay UI
+
+For simple scenarios, we provide a UI Overlay that can be used to spawn bots in the scene.
+This allows you to quickly try out automations without having to write code to manage the bots.
+To add the Overlay UI, click **GameObject > Regression Games > UI Overlay** in the Unity menu.
+
+![The Overlay UI running in the Boss Room sample](img/getting-started/overlay-menu.png)
+
+The UI Overlay provides a small button in the bottom-right corner of the game.
+When clicked, this button will open a dialog that shows you a list of available bots that can be spawned.
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md:83]: Yeah, because they're separate concerns. The automation _engine_ doesn't depend on the Overlay UI. There are APIs on Automation Controller to start/stop bots and handle recordings. The Overlay UI is purely a UI for interacting with those APIs without having to write custom code. I could see a production game having the Automation Controller but **not** having the UI Overlay.

**Comment by analogrelay** [comment]
PR Feedback addressed! Taking this out of draft as the remaining work is to record a GIF of the bot and I don't want to block people reviewing the doc on just that. I'll work on that right after lunch though.

**Comment by analogrelay** [comment]
And now the GIF recording is in place!

**Review by vontell — COMMENTED** [review]
Just cleaning up the addressed comments, I believe there are some that are still waiting to be addressed?

File: docs/automated-testing/getting-started.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+Import it into your game by clicking **Window > Package Manager** in the Unity menu,
+and then clicking the **+** button in the top left corner of the Package Manager window.
+Select **Add package from git URL** and paste in the following URL:
+
+```
+https://github.com/Regression-Games/RegressionGames.Unity.Testing.git?path=src/gg.regression.unity.testing#v0.0.1
+```
+
+* **NOTE**: The package uses [TextMeshPro](https://docs.unity3d.com/Manual/com.unity.textmeshpro.html) to provide UI elements. If you are prompted by Unity to add TextMeshPro assets to your project, please add them.
+* **NOTE**: The package uses [Newtonsoft Json](https://docs.unity3d.com/Packages/com.unity.nuget.newtonsoft-json@3.1/manual/index.html) to serialize state for recordings.
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md]: Done!

File: docs/automated-testing/getting-started.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+**By default**, the Automation Controller and all it's children will mark themselves as [DontDestroyOnLoad](https://docs.unity3d.com/ScriptReference/Object.DontDestroyOnLoad.html). This ensures that bots continue to run when you load a new scene in your game. You can disable this behavior by unchecking the "Dont Destroy On Load" checkbox on the Automation Controller component.
+
+![The Automation Controller component with the "Dont Destroy On Load" checkbox checked](img/getting-started/ac-dont-destroy-on-load.png)
+
+## Adding Entity Discoverers to your scene
+
+In order to automate entities in your game, the Automation Controller needs to be able to find them.
+The Automation Controller uses Entity Discoverers to find entities in the scene that can be automated.
+The package provides a few built-in Entity Discoverers, but you can also create your own.
+To add one of the built-in discoverers, open the **GameObject > Regression Games > Discovery** menu and select the discoverer you want to add:
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md]: > Here, also be sure to add that you should do this on the automation controller, unless it can be done at the top level as well (but we should at least tell them somewhere to click to get that menu)

If you click that menu item without a node selected, it automatically targets the automation controller. Otherwise, it should give you a fairly clear error telling you the problem.

> I see... this comes after the picture. I'd put it before the picture so that we avoid the scenario I just had where I didn't scroll further to see more and got stuck on that

I didn't want to put it the image above since the text that describes the image should generally precede it. I can do that if you think it's clearer though.

File: docs/automated-testing/getting-started.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+![The "Available Bots" property on the UI Overlay component](img/getting-started/overlay-available-bots.png)
+
+The UI Overlay includes the built-in "Monkey Bot", which is a bot that randomly clicks on UI elements in your game and records a screenshot of the game after each click.
+
+**By default**, the Overlay UI will mark itself as [DontDestroyOnLoad](https://docs.unity3d.com/ScriptReference/Object.DontDestroyOnLoad.html). This allows you to retain control over starting and stopping bots when you load a new scene in your game. You can disable this behavior by unchecking the "Dont Destroy On Load" checkbox on the Overlay UI component.
+
+![The Automation Controller component with the "Dont Destroy On Load" checkbox checked](img/getting-started/ac-dont-destroy-on-load.png)
+
+## Running the Monkey Bot
```
[the remark this answers is not in the document]
Comment (analogrelay) [reply docs/automated-testing/getting-started.md]: Updated!

**Comment by analogrelay** [comment]
> Just cleaning up the addressed comments, I believe there are some that are still waiting to be addressed?

I thought I got them all, but feel free to let me know if I missed something!

File: docs/automated-testing/getting-started.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/automated-testing/getting-started.md]: Good point! I like your rewrite.

**Review by abeizer — APPROVED** [review]

File: docs/automated-testing/getting-started.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (vontell) [reply docs/automated-testing/getting-started.md]: I think this change still needs to be applied?

**Review by addisonbgross — APPROVED** [review]
This reads really well, nice one!
